src/main/capi/arrow-c.cpp

```diff
@@ -14,6 +16,154 @@ using duckdb::PreparedStatementWrapper;
+	} catch (...) {
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT, "Unknown error occurred during conversion");
+	}
+	return nullptr;
+}
+
+duckdb_error_data arrow_to_duckdb_schema(duckdb_connection connection, struct ArrowSchema *schema,
+                                         duckdb_arrow_converted_schema *out_types, char ***out_names,
+                                         idx_t *out_column_count) {
+	if (!connection || !out_types || !out_names || !out_column_count || !schema) {
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT,
+		                                "Invalid argument(s) to duckdb_data_chunk_to_arrow");
+	}
+	duckdb::vector<std::string> names;
+	auto conn = reinterpret_cast<Connection *>(connection);
+	auto arrow_table = new duckdb::ArrowTableType();
+	*out_types = reinterpret_cast<duckdb_arrow_converted_schema>(arrow_table);
+	try {
+		duckdb::vector<LogicalType> return_types;
+		duckdb::ArrowTableFunction::PopulateArrowTableType(duckdb::DBConfig::GetConfig(*conn->context), *arrow_table,
+		                                                   *schema, names, return_types);
+		QueryResult::DeduplicateColumns(names);
+	} catch (const duckdb::Exception &ex) {
+		delete arrow_table;
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT, ex.what());
+	} catch (const std::exception &ex) {
+		delete arrow_table;
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT, ex.what());
+	} catch (...) {
+		delete arrow_table;
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT, "Unknown error occurred during conversion");
+	}
+	const idx_t column_count = names.size();
+	*out_column_count = column_count;
+	*out_names = new char *[column_count + 1];
+	for (idx_t i = 0; i < column_count; i++) {
+		(*out_names)[i] = new char[names[i].size() + 1];
+		std::strcpy((*out_names)[i], names[i].c_str());
+	}
+	// null-terminate the list
+	(*out_names)[column_count] = nullptr;
+	return nullptr;
+}
+
+duckdb_error_data arrow_to_duckdb_data_chunk(duckdb_connection connection, struct ArrowArray *arrow_array,
+                                             duckdb_arrow_converted_schema converted_schema,
+                                             duckdb_data_chunk *out_chunk) {
+	if (!connection || !converted_schema || !out_chunk || !arrow_array) {
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT,
+		                                "Invalid argument(s) to duckdb_data_chunk_to_arrow");
+	}
+	auto arrow_table = reinterpret_cast<duckdb::ArrowTableType *>(converted_schema);
+	auto conn = reinterpret_cast<Connection *>(connection);
+	auto &type_vector = arrow_table->GetTypes();
+	std::vector<duckdb::LogicalType *> types_ptr(type_vector.size());
+	for (idx_t i = 0; i < type_vector.size(); i++) {
+		types_ptr[i] = &type_vector[i];
+	}
+
+	*out_chunk = duckdb_create_data_chunk(reinterpret_cast<duckdb_logical_type *>(&types_ptr[0]), types_ptr.size());
```

Comment:
Instead of jumping through hoops to call another C API method we can directly construct the data chunk, then we can also immediately pass in the capacity:

```
auto result = duckdb::make_uniq<duckdb::DataChunk>();
result->Initialize(duckdb::Allocator::DefaultAllocator(), types, arrow_array->length);
*out_chunk = reinterpret_cast<duckdb_data_chunk>(result.release());
```